src/probabilities_estimators/permutation_ordinal/SymbolicPermutation.jl

```diff
@@ -76,74 +67,79 @@ information about within-state-vector amplitudes.
-    Permutation entropy based time series analysis: Equalities in the input signal can
+    Permutation entropy based timeseries analysis: Equalities in the input signal can
     lead to false conclusions. Physics Letters A, 381(22), 1883-1892.
 [^He2016]:
     He, S., Sun, K., & Wang, H. (2016). Multivariate permutation entropy and its
     application for complexity analysis of chaotic systems. Physica A: Statistical
     Mechanics and its Applications, 461, 812-823.
 """
-struct SymbolicPermutation{F} <: PermutationProbabilitiesEstimator
+struct SymbolicPermutation{M,F} <: PermutationProbabilitiesEstimator
+    encoding::OrdinalPatternEncoding{M,F}
     τ::Int
-    m::Int
-    lt::F
 end
-function SymbolicPermutation(; τ::Int = 1, m::Int = 3, lt::F=isless_rand) where {F <: Function}
-    m >= 2 || throw(ArgumentError("Need m ≥ 2, otherwise no dynamical information is encoded in the symbols."))
-    SymbolicPermutation{F}(τ, m, lt)
+function SymbolicPermutation(; τ::Int = 1, m::Int = 3, lt::F=isless_rand) where {F}
+    m >= 2 || throw(ArgumentError("Need order m ≥ 2."))
+    return SymbolicPermutation{m, F}(OrdinalPatternEncoding{m, F}(m, lt), τ)
 end
 
-function probabilities(est::SymbolicPermutation, x::Vector_or_Dataset)
-    probabilities(encodings_from_permutations(est, x))
+# Probabilities etc. simply initialize a the datasets and containers of the encodings
+# and just map everythihng using `encode`.
+function probabilities(est::SymbolicPermutation{m}, x::Vector{<:Real}) where {m}
+    dataset = genembed(x, m, est.τ)
+    return probabilities(est, dataset)
 end
 
-function probabilities!(πs::AbstractVector{Int}, est::SymbolicPermutation, x::Vector_or_Dataset)
-    encodings_from_permutations!(πs, est, x)
-    probabilities(πs)
+function probabilities(est::SymbolicPermutation{m}, x::AbstractDataset{D}) where {m, D}
+    m != D && throw(ArgumentError(
+        "Order of ordinal patterns and dimension of `Dataset` must match!"
+    ))
+    πs = zeros(Int, length(x))
+    return probabilities!(πs, est, x)
 end
 
-function probabilities_and_outcomes(est::SymbolicPermutation, x::Vector_or_Dataset)
-    πs = encodings_from_permutations(est, x)
-    return probabilities(πs), observed_outcomes(est, πs)
+function probabilities!(::Vector{Int}, ::SymbolicPermutation, ::AbstractVector)
+    error("""
+    In-place `probabilities!` for `SymbolicPermutation` can only be used by
+    Dataset input, not timeseries. First embed the timeseries or use the
+    normal version `probabilities`.
+    """)
 end
 
-function probabilities_and_outcomes!(πs::AbstractVector{Int}, est::SymbolicPermutation, x::Vector_or_Dataset)
-    encodings_from_permutations!(πs, est, x)
-    return probabilities(πs), observed_outcomes(est, πs)
+function _probabilities!(πs::Vector{Int}, est::SymbolicPermutation{m}, x::AbstractDataset{m}) where {m}
+    # TODO: The following loop can probably be parallelized!
```

Comment:
I was thinking about that too yesterday while doing some practical examples, where I ended up using `Distributed`  arrays and `@sync` `@distributed` on the for loop. However, maybe its possible to use `@threads`?

I haven't done any serious parallelization since Julia 0.5 or something like that, so I'm not up to speed on current best practice.

Reply:
Anyhow: the TODO should be an issue, so we'll remember to actually address it, or dismiss the idea, if that's what we want.

Reply:
Probably best to use https://github.com/JuliaSIMD/Polyester.jl  because its loop is rather cheap. I will think about this once the API is super stable and 2.0 is out.

Reply:
Some of the methods are also simple enough that one could possibly paralellize for the GPU too. That would be incredibly useful for the spatial methods, e.g. in examples like in #215 (where I had to precompute entropies for the animations, because it's too expensive to do in real time) .